File: backend/users/webauthn.py
from django.conf import settings
from django.http import HttpRequest

from webauthn import (
    generate_authentication_options,
    generate_registration_options,
    verify_authentication_response,
    verify_registration_response,
)
from webauthn.helpers.structs import (
    AuthenticatorSelectionCriteria,
    PublicKeyCredentialDescriptor,
    ResidentKeyRequirement,
    UserVerificationRequirement,
)


def get_rp_id_for_request(request: HttpRequest) -> str:
    """Derive the WebAuthn RP ID from the request host, validated against allowed RP IDs."""
    host = request.get_host().split(":")[0]  # strip port
    if host in settings.WEBAUTHN_RP_ID:
        return host
    # Fallback to first configured RP ID
    return settings.WEBAUTHN_RP_ID[0]


def get_registration_options(
    user_id: str, user_email: str, existing_credentials: list[bytes], rp_id: str
):
    options = generate_registration_options(
        rp_id=rp_id,
        rp_name=settings.WEBAUTHN_RP_NAME,
        user_id=user_id.encode(),
        user_name=user_email,
        authenticator_selection=AuthenticatorSelectionCriteria(
            resident_key=ResidentKeyRequirement.REQUIRED,
            user_verification=UserVerificationRequirement.REQUIRED,
        ),
        exclude_credentials=[
            PublicKeyCredentialDescriptor(id=cred_id) for cred_id in existing_credentials
        ],
    )
    return options


def verify_registration(credential_json: str, challenge: bytes, rp_id: str):
    return verify_registration_response(
        credential=credential_json,
        expected_challenge=challenge,
        expected_rp_id=rp_id,
        expected_origin=settings.WEBAUTHN_ORIGIN,
    )


def get_authentication_options(credential_ids: list[bytes], rp_id: str):
    options = generate_authentication_options(
        rp_id=rp_id,
        allow_credentials=[PublicKeyCredentialDescriptor(id=cred_id) for cred_id in credential_ids],
        user_verification=UserVerificationRequirement.REQUIRED,
    )
    return options


def verify_authentication(
    credential_json: str,
    challenge: bytes,
    credential_public_key: bytes,
    credential_current_sign_count: int,
    rp_id: str,
):
    return verify_authentication_response(
        credential=credential_json,
        expected_challenge=challenge,
        expected_rp_id=rp_id,
        expected_origin=settings.WEBAUTHN_ORIGIN,
        credential_public_key=credential_public_key,
        credential_current_sign_count=credential_current_sign_count,
    )

File: web/lib/auth/webauthn.ts
import {
  generateAuthenticationOptions,
  generateRegistrationOptions,
  verifyAuthenticationResponse,
  verifyRegistrationResponse,
  type PublicKeyCredentialCreationOptionsJSON,
  type PublicKeyCredentialRequestOptionsJSON,
} from "@simplewebauthn/server";
import { getAllowedOrigins, getRpName } from "./config";
import { AuthError } from "./errors";

export function bufToB64url(buf: Buffer): string {
  return buf.toString("base64url");
}

export function b64urlToBuf(s: string): Buffer {
  return Buffer.from(s, "base64url");
}

export async function getRegistrationOptions(args: {
  userId: string;
  userEmail: string;
  rpId: string;
  excludeCredentialIds: Buffer[];
}): Promise<PublicKeyCredentialCreationOptionsJSON> {
  return generateRegistrationOptions({
    rpName: getRpName(),
    rpID: args.rpId,
    userID: new TextEncoder().encode(args.userId),
    userName: args.userEmail,
    attestationType: "none",
    excludeCredentials: args.excludeCredentialIds.map((id) => ({ id: bufToB64url(id) })),
    authenticatorSelection: {
      residentKey: "required",
      userVerification: "required",
    },
  });
}

export async function verifyRegistration(args: {
  responseJson: string;
  expectedChallenge: string;
  rpId: string;
}): Promise<{ credentialId: Buffer; publicKey: Buffer; signCount: number }> {
  let verification;
  try {
    verification = await verifyRegistrationResponse({
      response: JSON.parse(args.responseJson),
      expectedChallenge: args.expectedChallenge,
      expectedOrigin: getAllowedOrigins(),
      expectedRPID: args.rpId,
      requireUserVerification: true,
    });
  } catch (e) {
    throw new AuthError(400, `WebAuthn verification failed: ${(e as Error).message}`);
  }
  if (!verification.verified || !verification.registrationInfo) {
    throw new AuthError(400, "WebAuthn registration could not be verified.");
  }
  const cred = verification.registrationInfo.credential;
  return {
    credentialId: b64urlToBuf(cred.id),
    publicKey: Buffer.from(cred.publicKey),
    signCount: cred.counter,
  };
}

export async function getAuthenticationOptions(args: {
  rpId: string;
  allowCredentialIds: Buffer[];
}): Promise<PublicKeyCredentialRequestOptionsJSON> {
  return generateAuthenticationOptions({
    rpID: args.rpId,
    allowCredentials: args.allowCredentialIds.map((id) => ({ id: bufToB64url(id) })),
    userVerification: "required",
  });
}

export async function verifyAuthentication(args: {
  responseJson: string;
  expectedChallenge: string;
  rpId: string;
  credentialId: Buffer;
  publicKey: Buffer;
  signCount: number;
}): Promise<{ newSignCount: number }> {
  let verification;
  try {
    verification = await verifyAuthenticationResponse({
      response: JSON.parse(args.responseJson),
      expectedChallenge: args.expectedChallenge,
      expectedOrigin: getAllowedOrigins(),
      expectedRPID: args.rpId,
      requireUserVerification: true,
      credential: {
        id: bufToB64url(args.credentialId),
        publicKey: new Uint8Array(args.publicKey),
        counter: args.signCount,
      },
    });
  } catch (e) {
    throw new AuthError(400, `WebAuthn verification failed: ${(e as Error).message}`);
  }
  if (!verification.verified) {
    throw new AuthError(400, "WebAuthn authentication could not be verified.");
  }
  return { newSignCount: verification.authenticationInfo.newCounter };
}
